File: src/dataset/python/complex-numbers_exercism-ThomasZumsteg/gemini/GeminiFlash_complex_numbers_v1.py
import math

class ComplexNumber(object):
    def __init__(self, real, imaginary):
        self.real = real
        self.imaginary = imaginary

    def __add__(self, other):
        return ComplexNumber(self.real + other.real, self.imaginary + other.imaginary)

    def __mul__(self, other):
        real = self.real * other.real - self.imaginary * other.imaginary
        imaginary = self.real * other.imaginary + self.imaginary * other.real
        return ComplexNumber(real, imaginary)

    def __sub__(self, other):
        return ComplexNumber(self.real - other.real, self.imaginary - other.imaginary)

    def __truediv__(self, other):
        common = other.real**2 + other.imaginary**2
        real = (self.real * other.real + self.imaginary * other.imaginary) / common
        imaginary = (self.imaginary * other.real - self.real * other.imaginary) / common
        return ComplexNumber(real, imaginary)

    def __abs__(self):
        return math.sqrt(self.real**2 + self.imaginary**2)

    def conjugate(self):
        return ComplexNumber(self.real, -self.imaginary)

    def exp(self):
        exp_real = math.exp(self.real)
        real = exp_real * math.cos(self.imaginary)
        imaginary = exp_real * math.sin(self.imaginary)
        return ComplexNumber(real, imaginary)

    def __eq__(self, other):
        return self.real == other.real and self.imaginary == other.imaginary

    def __repr__(self):
        return "{} + {}i".format(self.real, self.imaginary)

File: src/dataset/javascript/complex-numbers_exercism-ThomasZumsteg/gemini/GeminiFlash_complex-numbers_v1.js
class ComplexNumber {
    constructor(real, imag) {
        this.real = real;
        this.imag = imag;
    }

    add(other) {
        return new ComplexNumber(this.real + other.real, this.imag + other.imag);
    }

    sub(other) {
        return new ComplexNumber(this.real - other.real, this.imag - other.imag);
    }

    mul(other) {
        const real = this.real * other.real - this.imag * other.imag;
        const imag = this.real * other.imag + this.imag * other.real;
        return new ComplexNumber(real, imag);
    }

    div(other) {
        const otherReal = other.real;
        const otherImag = other.imag;
        const denom = otherReal * otherReal + otherImag * otherImag;

        if (denom === 0) {
            return new ComplexNumber(NaN, NaN);
        }

        const realNum = this.real * otherReal + this.imag * otherImag;
        const imagNum = this.imag * otherReal - this.real * otherImag;
        return new ComplexNumber(realNum / denom, imagNum / denom);
    }

    get abs() {
        const real = this.real;
        const imag = this.imag;
        return Math.sqrt(real * real + imag * imag);
    }

    get conj() {
        return new ComplexNumber(this.real, -this.imag);
    }

    get exp() {
        const expReal = Math.exp(this.real);
        const cosImag = Math.cos(this.imag);
        const sinImag = Math.sin(this.imag);
        return new ComplexNumber(expReal * cosImag, expReal * sinImag);
    }
}

module.exports = ComplexNumber;